src/rgw/driver/d4n/d4n_datacache.cc

```diff
@@ -0,0 +1,482 @@
+	  if (reply.is_array()) {
+	    auto arr = reply.as_array();
+
+	    if (!arr[0].is_null()) {
+	      key_exist = 0;
+
+	      for (long unsigned int i = 0; i < getFields.size(); ++i) {
+		newMetadata->push_back({getFields[i], arr[i].as_string()});
+	      }
+	    }
+	  }
+	});
+
+	client.sync_commit(std::chrono::milliseconds(1000));
+      } catch(std::exception &e) {
+	return -1;
+      }
+    }
+  }
+
+  if (key_exist < 0) {
+    dout(20) << "RGW D4N Cache: Object was not retrievable." << dendl;
+  }
+
+  return key_exist;
+}
+
+int RGWD4NCache::copyObject(std::string original_oid, std::string copy_oid, rgw::sal::Attrs* attrs) {
+  std::string result;
+  std::vector< std::pair<std::string, std::string> > redisObject;
+  std::string key = "rgw-object:" + original_oid + ":cache";
+
+  if (!client.is_connected()) {
+    findClient(&client);
+  }
+
+  /* Read values from cache */
+  if (existKey(key)) {
+    try {
+      client.hgetall(key, [&redisObject](cpp_redis::reply &reply) {
+        if (reply.is_array()) {
+	  auto arr = reply.as_array();
+
+	  if (!arr[0].is_null()) {
+            for (long unsigned int i = 0; i < arr.size() - 1; i += 2) {
+	      redisObject.push_back({arr[i].as_string(), arr[i + 1].as_string()});
+	    }
+	  }
+	}
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+    } catch(std::exception &e) {
+      return -1;
+    }
+  } else {
+    return -2; 
+  }
+
+  /* Build copy with updated values */
+  if (!redisObject.empty()) {
+    rgw::sal::Attrs::iterator attr;
+    
+    for (attr = attrs->begin(); attr != attrs->end(); ++attr) {
+      auto it = std::find_if(redisObject.begin(), redisObject.end(),
+        [&](const auto& pair) { return pair.first == attr->first; });
+
+      if (it != redisObject.end()) {
+	int index = std::distance(redisObject.begin(), it);
+	redisObject[index] = {attr->first, attr->second.to_str()};
+      } else {
+	redisObject.push_back(std::make_pair(attr->first, attr->second.to_str()));
+      }
+    }
+  }
+
+  /* Set copy with new values */
+  key = "rgw-object:" + copy_oid + ":cache";
+
+  try {
+    client.hmset(key, redisObject, [&result](cpp_redis::reply &reply) {
+      if (!reply.is_null()) {
+        result = reply.as_string();
+      }
+    });
+
+    client.sync_commit(std::chrono::milliseconds(1000));
+
+    if (result != "OK") {
+      return -1;
+    }
+  } catch(std::exception &e) {
+    return -2;
+  }
+   
+  return 0;
+}
+
+int RGWD4NCache::delObject(std::string oid) {
+  int result = 0;
+  std::vector<std::string> keys;
+  std::string key = "rgw-object:" + oid + ":cache";
+  keys.push_back(key);
+
+  if (!client.is_connected()) {
+    findClient(&client);
+  }
+
+  if (existKey(key)) {
+    try {
+      client.del(keys, [&result](cpp_redis::reply &reply) {
+	if (reply.is_integer()) {
+	  result = reply.as_integer();
+	}
+      });
+
+      client.sync_commit(std::chrono::milliseconds(1000));
+      
+      return result - 1;
```

Comment:
@samarahu what are the expected ranges of values here for `result`? Why not just return a positive or negative constant like `-1` or `-2`?

Reply:
Since deletes return a 1 upon success, returning the `result - 1` eliminates the need to check for the value of the result and map it to the values that the SAL layer uses (0 is for success rather than 1). I could also just check if the return value of these operations is equal to 1 but I wanted to stay consistent with the SAL layer. It's very minor but otherwise, I'd need to add an if statement.

Reply:
Got it. FYI after talking with the group, the return values we care about are the ones in rgw_sal_d4n. None of the return values in the sal layer return values from `RGWD4NCache` or `RGWBlockDirectory calls.